### PR TITLE
NTFS Reparse Point proof-of-concept

### PR DESCRIPTION
**Pull request by NTmatter**
Lightly-tested modifications for read-only scenario in #1845. Deduplicated files seem to be correctly scanned and synced.

Note, this has not been tested with a read-write folder, so odds are that things will break when attempting to overwrite something that is still incorrectly perceived as a symlink.

### CODE REVIEW

**Comment by Zillode** [comment]
@st-jenkins run integration tests for windows please

**Comment by calmh** [comment]
@st-jenkins add to whitelist

File: internal/symlinks/symlink_windows.go [hunk 1]
```diff
@@ -125,6 +127,11 @@ func Read(path string) (string, uint32, error) {
 		flags = protocol.FlagSymlinkMissingTarget
 	} else if attr&syscall.FILE_ATTRIBUTE_DIRECTORY != 0 {
 		flags = protocol.FlagDirectory
+		// TODO Check mountpoint reparse point to see if Directory is set.
+	} else if data.reparseTag == IO_REPARSE_TAG_DEDUP {
+		// Deduplicated files are effectively normal files, except that they
+		// must be Lstat'ed rather than Stat'ed.
+		return osutil.NormalizedFilename(path), flags, nil
```
Comment (calmh) [inline internal/symlinks/symlink_windows.go]: Returning an explicit `0` instead of `flags` here would be more clear I think

**Comment by calmh** [comment]
Looks sane to me. Not sure about the test failures, doesn't look related to this change as far as I can tell, will look into it. `stfileinfo/main.go` needs gofmt.

File: internal/scanner/walk.go [hunk 2]
```diff
@@ -117,6 +117,18 @@ func (w *Walker) walkAndHashFiles(fchan chan protocol.FileInfo) filepath.WalkFun
 		// Return value used when we are returning early and don't want to
 		// process the item. For directories, this means do-not-descend.
 		var skip error // nil
+
```
Comment (calmh) [inline internal/scanner/walk.go]: Ah. This section needs to be added after the error checks on (current) line 137-142, or `info` may be nil and this explodes.

**Comment by calmh** [comment]
Probably the commits should be squashed into a single one like "Correctly handle deduplicated files on NTFS (fixes #1845)" (with the fixes suffix being essential, the rest whatever sounds reasonable).

**Comment by calmh** [comment]
I'm assuming that creating normal files works as usual on this type of filesystem, and that deduplication is either automatic and transparent or a background process of some kind?

**Comment by NTmatter** [comment]
Yes, files are created and accessed as normal.

I still want to take a slightly closer look at the process of *replacing* a deduplicated file with a new regular one (or with a symlink), just to ensure that the write path behaves itself. Notably, I'm not sure if [rwfolder.handleFile](https://github.com/syncthing/syncthing/blob/65be18cc93aeca4c60cd87a1228c5593e4b908db/internal/model/rwfolder.go#L780) needs to be modified to cope with the potentially inaccurate Symlink flag.

As for Deduplication, it's a background process that is mostly transparent. While the dedupe engine is running, it seems to acquire a write lock on the file(s) that it is thinning, which can temporarily hold up modification of very large files. Aside from the temporary lock and conversion to a reparse point, file attributes and contents are otherwise unchanged.

**Comment by calmh** [comment]
Any chance of getting this updated?

**Comment by NTmatter** [comment]
Yep, this can probably be closed as Invalid/WontFix for now. I'm looking at submitting some patches to Go itself which should (hopefully) smooth out the issue without requiring any changes here.

**Comment by calmh** [comment]
OK. Be back with a patch here otherwise. :)

**Pull request closed without merging**